hal/spi_api.h

```diff
@@ -19,220 +18,259 @@
+ * SS      \_______________/ -- \_______________/ -- \_______________/ -- \_______________/
+ *        _ ___         _____  _____         _____  _         _________  _____             __
+ * DATA   _X   \_______/   X_--_X   \_______/   X_--_X_______/       X_--_X   \___________X__
+ * _
+ * ```
+ * ## Symbol/Word
+ * Even though documentations usually uses `word` to mean the unit of data that is sent/received
+ * it is ambiguous as it may not correspond to an actual `word` on the platform. To avoid any
+ * confusion we will use in this API the word `symbol` instead.
+ * ## SPI channel
+ * This is the virtual communication channel created between a master and its slave.
+ * @{
+ */
 
-#define SPI_EVENT_ERROR       (1 << 1)
-#define SPI_EVENT_COMPLETE    (1 << 2)
-#define SPI_EVENT_RX_OVERFLOW (1 << 3)
-#define SPI_EVENT_ALL         (SPI_EVENT_ERROR | SPI_EVENT_COMPLETE | SPI_EVENT_RX_OVERFLOW)
-
-#define SPI_EVENT_INTERNAL_TRANSFER_COMPLETE (1 << 30) // Internal flag to report that an event occurred
-
-#define SPI_FILL_WORD         (0xFFFF)
-#define SPI_FILL_CHAR         (0xFF)
+/**
+ * This is the hal level SPI "class".
+ *
+ * The actual definition of this structure is delegated to the device implementation of the hal.
+ */
+typedef struct spi_t spi_t;
 
-#if DEVICE_SPI_ASYNCH
-/** Asynch SPI HAL structure
+/**
+ * This holds together the pins related to a single SPI channel.
+ *
+ * Any pin but the Slave Select pin may be shared between any number of SPI channels.
+ * The Slave Select pin serves as a "unique identifier".
  */
 typedef struct {
-    struct spi_s spi;        /**< Target specific SPI structure */
-    struct buffer_s tx_buff; /**< Tx buffer */
-    struct buffer_s rx_buff; /**< Rx buffer */
-} spi_t;
+    PinName ss; /**< Slave select pin. */
+    PinName miso; /**< Master In Slave Out. */
+    PinName mosi; /**< Master Out Slave In. */
+    PinName mclk; /**< Master Clock pin. */
+} spi_pins_t;
 
-#else
-/** Non-asynch SPI HAL structure
+/**
+ * This structure groups all initialisation parameters required by a SPI interface.
  */
-typedef struct spi_s spi_t;
+typedef struct {
+    spi_pins_t pins; /**< Pins to use for this channel. */
 
-#endif
+    /**< This symbol will be sent as a place holder in full-duplex mode in case more symbols are to
+         be read than to be sent. */
+    uint32_t fill_symbol;
+    uint32_t clock_frequency; /**< MCLK frequency in Hz. */
 
-#ifdef __cplusplus
-extern "C" {
-#endif
+    uint32_t word_length; /**< Length of a symbol in bit. */
+    bool is_master; /**< True to configure the device in Master mode. */
+    bool msb_first; /**< True to send/receive the most significant bit first. */
+    bool clock_phase; /**< True if data line is valid when leaving active state. */
+    bool clock_polarity; /**< True if the clock's rest state is high (+Vcc). */
+    bool continuous_mode; /**< True to use the continuous mode. */
+
+    spi_mode_t  mode; /**< Transmission mode. See spi_mode_t. */
```

Comment:
The definition of the `spi_mode_t` type has been removed. It looks like `mode` field should be removed from `spi_init_t`.

Reply:
👍 Fixed